cirq-core/cirq/ops/pauli_string_phasor.py

```diff
@@ -58,138 +60,278 @@ def __init__(
+
+        if dense_pauli_string.coefficient != 1:
+            raise ValueError(
+                "Given PauliString doesn't have +1 and -1 eigenvalues. "
+                "pauli_string.coefficient must be 1 or -1."
+            )
+
+        self._dense_pauli_string = dense_pauli_string
+        self._exponent_neg = value.canonicalize_half_turns(exponent_neg)
+        self._exponent_pos = value.canonicalize_half_turns(exponent_pos)
+
+    @property
+    def exponent_relative(self) -> Union[int, float, sympy.Basic]:
+        return value.canonicalize_half_turns(self._exponent_neg - self._exponent_pos)
+
+    @property
+    def exponent_neg(self):
+        return self._exponent_neg
+
+    @property
+    def exponent_pos(self):
+        return self._exponent_pos
+
+    @property
+    def dense_pauli_string(self):
+        return self._dense_pauli_string
+
+    def _value_equality_values_(self):
+        return (
+            self._dense_pauli_string,
+            self._exponent_neg,
+            self._exponent_pos,
+        )
+
+    def equal_up_to_global_phase(self, other):
+        if isinstance(other, PauliStringPhasorGate):
+            rel1 = self.exponent_relative
+            rel2 = other.exponent_relative
+            return rel1 == rel2 and self._dense_pauli_string == other.dense_pauli_string
+        return False
+
+    def __pow__(self, exponent: Union[float, sympy.Symbol]) -> 'PauliStringPhasorGate':
+        pn = protocols.mul(self._exponent_neg, exponent, None)
+        pp = protocols.mul(self._exponent_pos, exponent, None)
+        if pn is None or pp is None:
+            return NotImplemented
+        return PauliStringPhasorGate(self._dense_pauli_string, exponent_neg=pn, exponent_pos=pp)
 
     def _has_unitary_(self):
         return not self._is_parameterized_()
 
-    def _decompose_(self) -> 'cirq.OP_TREE':
-        if len(self.pauli_string) <= 0:
+    def _to_z_basis_ops(self, qubits: Sequence['cirq.Qid']) -> Iterator[raw_types.Operation]:
+        """Returns operations to convert the qubits to the computational basis."""
+        for i in range(len(self._dense_pauli_string)):
+            yield clifford_gate.SingleQubitCliffordGate.from_single_map(
+                {self._dense_pauli_string[i]: (pauli_gates.Z, False)}
+            )(qubits[i])
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        if len(self._dense_pauli_string) <= 0:
             return
-        qubits = self.qubits
         any_qubit = qubits[0]
-        to_z_ops = op_tree.freeze_op_tree(self.pauli_string.to_z_basis_ops())
+        to_z_ops = op_tree.freeze_op_tree(self._to_z_basis_ops(qubits))
         xor_decomp = tuple(xor_nonlocal_decompose(qubits, any_qubit))
         yield to_z_ops
         yield xor_decomp
 
-        if self.exponent_neg:
-            yield pauli_gates.Z(any_qubit) ** self.exponent_neg
-        if self.exponent_pos:
+        if self._exponent_neg:
```

Comment:
Use `self.exponent_neg` property instead of the internal variable `self._exponent_neg`. Here and elsewhere for all properties.